File: packages/rails/test/task_test.rb
# frozen_string_literal: true

require "json"
require "minitest/autorun"
require_relative "../lib/pi/browser/taskbar/rails/task"

class RailsTaskTest < Minitest::Test
  FIXTURE = File.expand_path("../../../contract/fixtures/tasks/minimal-task.json", __dir__)
  GOLDEN = File.expand_path("../../../contract/fixtures/prompts/minimal-task.json", __dir__)

  def test_validates_and_builds_the_canonical_trusted_untrusted_prompt
    value = JSON.parse(File.read(FIXTURE))
    task = Pi::Browser::Taskbar::Rails::Task.parse(value)
    golden = JSON.parse(File.read(GOLDEN))

    assert_equal value["prompt"], task.prompt
    assert_equal golden.fetch("expected"), task.pi_prompt
  end

  def test_rejects_unknown_fields_and_normalizes_prompt
    value = JSON.parse(File.read(FIXTURE))
    value["unknown"] = true
    error = assert_raises(Pi::Browser::Taskbar::Rails::Task::Invalid) { Pi::Browser::Taskbar::Rails::Task.parse(value) }
    assert_match(/unknown field/, error.message)

    value.delete("unknown")
    value["prompt"] = " \r\nExplain\u202ethe page. \r"
    assert_equal "Explainthe page.", Pi::Browser::Taskbar::Rails::Task.parse(value).prompt
  end

  def test_rejects_oversized_prompt_and_snapshot_depth
    value = JSON.parse(File.read(FIXTURE))
    value["prompt"] = "x" * 4_001
    assert_raises(Pi::Browser::Taskbar::Rails::Task::Invalid) { Pi::Browser::Taskbar::Rails::Task.parse(value) }

    value = JSON.parse(File.read(FIXTURE))
    node = value.dig("context", "snapshot")
    13.times { node["children"] = [{"tag" => "div", "children" => []}]; node = node["children"].first }
    assert_raises(Pi::Browser::Taskbar::Rails::Task::Invalid) { Pi::Browser::Taskbar::Rails::Task.parse(value) }
  end
end

File: packages/phoenix/test/task_test.exs
defmodule PiBrowserTaskbarPhoenix.TaskTest do
  use ExUnit.Case, async: true

  alias PiBrowserTaskbarPhoenix.Task

  @fixture Path.expand("../../../contract/fixtures/tasks/minimal-task.json", __DIR__)
  @golden Path.expand("../../../contract/fixtures/prompts/minimal-task.json", __DIR__)

  test "validates a whole-page request and separates instruction from untrusted context" do
    params = @fixture |> File.read!() |> Jason.decode!()
    golden = @golden |> File.read!() |> Jason.decode!()

    assert {:ok, task} = Task.new(params)
    assert task.prompt == "Explain the cards page."
    assert task.context["focus_points"] == []
    assert Task.to_prompt(task) == golden["expected"]
  end

  test "rejects unknown fields and empty normalized prompts" do
    params = @fixture |> File.read!() |> Jason.decode!()

    assert {:error, :invalid_task, message} = Task.new(Map.put(params, "unexpected", true))
    assert message =~ "unknown field"

    assert {:error, :invalid_task, "prompt is required"} =
             Task.new(Map.put(params, "prompt", " \r\n\t "))
  end

  test "rejects a page snapshot outside the native page bounds" do
    params = @fixture |> File.read!() |> Jason.decode!()
    node = %{"tag" => "div", "children" => []}

    too_many_nodes =
      put_in(params, ["context", "snapshot"], %{
        "tag" => "main",
        "children" => List.duplicate(node, 750)
      })

    assert {:error, :invalid_task, "snapshot must contain at most 750 nodes"} =
             Task.new(too_many_nodes)

    oversized_snapshot =
      put_in(params, ["context", "snapshot"], %{
        "tag" => "main",
        "children" =>
          Enum.map(1..50, fn _index ->
            %{"tag" => "p", "text" => String.duplicate("x", 1_000), "children" => []}
          end)
      })

    assert {:error, :invalid_task, "snapshot is too large"} = Task.new(oversized_snapshot)
  end
end
